Introduce inactivity_timeout (#222)
openid/sharedsignals: openid-sharedsignals-framework-1_0.md

    target: https://tools.ietf.org/html/rfc6749#section-4.4
    title: The OAuth 2.0 Authorization Framework - Client Credentials Grant

>   EXPIRES_IN:

Reviewer: you don't need to describe an RFC. You can just add its number and the kramdown tool will do the rest.

Author: I wasn't sure how to make it point to a specific appendix, hence followed the example of CLIENTCRED above. Both CLIENTCRED and EXPIRES_IN are from the same RFC, so the details remain the same.

Would you know the right syntax to make it point to the appendix instead of the whole RFC (https://tools.ietf.org/html/rfc6749#appendix-A.14).

Reviewer: I guess you aren't really referring to "CLIENTCRED", right? The Expires_in definition in OAuth is pretty small, so one way you can do this is by referring to it as "Section A.14 of  RFC6749 {{RFC6749}}".

Author: Done in a later commit as suggested

    to be transmitted. The Receiver MAY remove subjects added this way via the
    `remove_subject_endpoint`.

> stream_ttl

Reviewer 2: can we make it `stream_ttl_seconds`? will save people working with the api a documentation lookup

Author: Most of the other names in the this/similar spec do not seem to use a suffix, regardless of units. E.g. min_verification_interval

So we could keep it as stream_ttl for consistency.

> > OPTIONAL. The lifetime of a stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication in that duration.

Reviewer: "Receiver initiated" needs to be clarified, IMO. Does it differ for poll and push streams? Is the appropriate response code from the receiver to a push from a Tx sufficient?

Author: Clarified the definition of "receiver-initiated" communication.

  If the Transmitter decides to update the stream, it MUST send a Stream Updated Event to the Receiver as described in {{status}}.

>   If the Receiver calls any endpoint in the Event Stream Management API ({{management}}), the Transmitter must refresh the TTL of that particular stream.

Reviewer 3: ```suggestion
  If the Receiver calls any endpoint in the Event Stream Management API ({{management}}), the Transmitter MUST refresh the TTL of that particular stream.
```

Author: Done

>   If the Receiver calls any endpoint in the Event Stream Management API ({{management}}), the Transmitter must refresh the TTL of that particular stream.

Reviewer 3: We might want to be explicit that contact from the receiver does not automatically move the stream out of a paused/disabled state caused by a TTL timeout.

Author: Should we leave it up to the transmitter? It might complicate the spec if paused/disabled states were treated differently based on whether they were caused by a TTL timeout (Tx not allowed to automatically change the state), vs other causes like Rx requesting the pause (Tx allowed to do whatever).

  This is useful in multi-stream systems to identify the stream for human actors. The
  transmitter MAY truncate the string beyond an allowed max length.

stream_ttl

> > **Transmitter-Supplied**, OPTIONAL. The refreshable lifetime of the stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.

Reviewer: You might want to expand the acronym "TTL" somewhere in the description. You say it precisely in the first sentence, but you just don't say what "TTL" stands for. It'll help clarify the name even though most folks will know what it is.

Author: SG, done.

> > **Transmitter-Supplied**, OPTIONAL. The refreshable lifetime of the stream in seconds, after which the Transmitter MAY either pause or disable the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.

Reviewer: Is the right behavior "pause or disable", or just "delete"?

Author: Pause / disable gives the Receiver an option to unpause / enable the stream, while delete means that the Receiver will have to recreate it.

I was imagining that some Transmitters may be more lenient where they simply pause the stream at first, then disable it after some time of being paused, and finally delete it if there's still no response (hence slowly escalating).

That said, either one works. Or we could even add the above paragraph as an example. What do you think?

Reviewer 3: Since it is MAY, you could say "...after which the Transmitter MAY either pause, disable, or delete the stream if it has not received..."

Author: Yeah we can keep the options open (implementation-dependent).

  This is useful in multi-stream systems to identify the stream for human actors. The
  transmitter MAY truncate the string beyond an allowed max length.

inactivity_timeout

> > **Transmitter-Supplied**, OPTIONAL. The refreshable inactivity timeout of the stream in seconds. After the timeout duration passes with no eligible activity from the Receiver, as defined below, the Transmitter MAY either pause, disable, or delete the stream. The syntax is the same as that of `expires_in` from Section A.14 of {{RFC6749}}.

Author: For PUSH streams that's not enough, at least from a security/privacy perspective - receiver endpoints are more likely to be open, the only way to verify that the receiver exists and still has the right auth/access is for it to call the transmitter endpoints (oauth-protected in most cases).

For PULL, polling is sufficient.

> > **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, since the last activity from the Receiver as described below, was conducted. After the TTL passes with no eligible activity from the Receiver, the Transmitter MAY either pause, disable, or delete the stream. The syntax is the same as that of `expires_in` from Section A.14 of {{RFC6749}}.

Reviewer 3: ```suggestion
> **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds. After the TTL passes with no eligible activity from the Receiver, as defined below, the Transmitter MAY either pause, disable, or delete the stream. The syntax is the same as that of `expires_in` from Section A.14 of {{RFC6749}}.
```
Nit suggestion for clarity

>

> > The following constitutes eligible Receiver activity, upon observing which, the Transmitter MUST reset its count toward the TTL of the stream to zero.

Reviewer 3: ```suggestion
> The following constitutes eligible Receiver activity. If the Transmitter observes any of these activities from the Receiver, it MUST restart the inactivity timeout counter.
```
Nit suggestion for clarity

  This is useful in multi-stream systems to identify the stream for human actors. The
  transmitter MAY truncate the string beyond an allowed max length.

inactivity_ttl

> > **Transmitter-Supplied**, OPTIONAL. The refreshable time-to-live (TTL) of the stream in seconds, after which the Transmitter MAY either pause, disable, or delete the stream if it has not received any Receiver-initiated communication (defined below) in that duration. The syntax is the same as that of {{EXPIRES_IN}}.

Reviewer: see suggestion above.

Author: Done

>

> > For streams created with the PUSH {{RFC8935}} delivery method, the Transmitter MUST refresh the TTL whenever:

Reviewer: "refresh the TTL" could be ambiguous. Can we be more precise? we could alter the description in line 1145 to read: "the refreshable time-to-live (TTL) of the stream in seconds, since the last activity from the Receiver as described below, was conducted. After the TTL, the Transmitter MAY..."

This sentence can then read: "Receiver activity can be any of:"

> > * The Receiver calls any endpoint in the Event Stream Management API ({{management}}).

Reviewer: We probably need to specify here and below that "calls to the Event Stream Management API that reference the stream"

Author: Done

>
> For streams created with the POLL {{RFC8936}} delivery method, the Transmitter MUST refresh the TTL whenever:
> * The Receiver polls the Transmitter for events.
> * The Receiver calls any endpoint in the Event Stream Management API ({{management}}).
>

> > If the Transmitter decides to update the stream, it MUST send a Stream Updated Event to the Receiver as described in {{status}}.

Reviewer 3: ```suggestion
> If the Transmitter decides to pause or disable the stream, it MUST send a Stream Updated Event to the Receiver as described in {{status}}.
```
It took me a second to understand why this sentence was here. Specifying pause or disable connects it to the language above.
